pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -450,17 +492,59 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 		// add one empty entry to the list so we generate a default listener below
 		allChains = []plugin.FilterChain{{}}
 	}
-	for _, chain := range allChains {
+
+	// Detect protocol by sniffing and double the filter chain
+	if pluginParams.ListenerProtocol == plugin.ListenerProtocolAuto {
+		allChains = append(allChains, allChains...)
+		listenerOpts.needHTTPInspector = true
+	}
+
+	for id, chain := range allChains {
```

Comment:
may be not in this PR but in another PR, can you refactor this code to not construct the httpinbound and tcp inbound filter stack again and again? You can construct them outside the loop once (based on the protocol) and then in the loop, you can assign the same object to different filter chains. Saves compute time and memory

Reply:
Yes. I planed to refactor this code after it's merged. I don't think we can assign the same object to different filter chains. I remember I fixed a bug before which multiple filter chain share one HCM object. But anyway we may find another way to refactor this code.